repair/repair.cc

```diff
@@ -1540,3 +1540,109 @@ future<> sync_data_with_repair(seastar::sharded<database>& db,
+                        // going to lose the ownership of the range, so there
+                        // is no need to choose a mandatory neighbor to sync
+                        // data from.
+                        // 2) If RF = old_endpoints.size(), it means one node is
+                        // going to lose the ownership of the range, we need to
+                        // choose it as the mandatory neighbor to sync data
+                        // from.
+                        std::vector<gms::inet_address> mandatory_neighbors;
+                        // All neighbors
+                        std::vector<inet_address> neighbors;
+                        if (old_endpoints.size() == strat.get_replication_factor()) {
+                            // Remove the new nodes from the old nodes list, so
+                            // that it contains only the node that will lose
+                            // the ownership of the range.
+                            mandatory_neighbors = boost::copy_range<std::vector<gms::inet_address>>(old_endpoints |
+                                    boost::adaptors::filtered([&new_endpoints] (const gms::inet_address& node) { return !new_endpoints.count(node); }));
+                            if (mandatory_neighbors.size() != 1) {
+                                throw std::runtime_error(format("bootstrap_with_repair: keyspace={}, range={}, expected 1 mandatory neighbor but found {}",
+                                            keyspace_name, desired_range, mandatory_neighbors.size()));
+                            }
+                            // For example, with RF = 3 and 3 nodes n1, n2, n3
+                            // in the cluster, n4 is bootstrapped, old_replicas
+                            // = {n1, n2, n3}, new_replicas = {n1, n2, n4}, n3
+                            // is losing the range. Choose the bootstrapping
+                            // node n4 to run repair to sync with the node
+                            // losing the range n3
+                            neighbors = mandatory_neighbors;
+                        } else if (old_endpoints.size() < strat.get_replication_factor()) {
+                            // For example, with RF = 3, 2 nodes n1, n2 in the
+                            // cluster, n3 is bootstrapped, old_replicas={n1, n2},
+                            // new_replicas={n1, n2, n3}. No node is losing
+                            // ranges. The bootstrapping node n3 has to sync
+                            // with n1 and n2, otherwise n3 might get the old
+                            // replica and make the total old replica be 2,
+                            // which makes the quorum read incorrect.
+                            // Choose the bootstrapping node n3 to reun repair
+                            // to sync with n1 and n2.
+                            for (auto& node : old_endpoints) {
```

Comment:
Please start a document in docs/ with this guiding information about repair-based node operations.